File: src/deadlock_build_sync/hero_cohort.py
"""Frozen per-hero rank ranges and their support-driven expansion history."""

from __future__ import annotations

from dataclasses import asdict, dataclass

from .artifacts import ArtifactError
from .build_evidence_values import _required_int
from .ranks import Rank, RankDivision, RankRange, RankTier
from .value_validation import object_dict, object_rows


def ranked_cutoffs(minimum: int, maximum: int, mode: str) -> tuple[int, ...]:
    if mode not in {"auto", "off"}:
        raise ValueError("rank expansion must be auto or off")
    RankRange(badge_rank(minimum), badge_rank(maximum))
    if mode == "off" or minimum == 11:
        return (minimum,)
    lower = tuple(range((minimum // 10 - 1) * 10 + 1, 10, -10))
    return (minimum, *(lower or (11,)))


def badge_rank(badge: int) -> Rank:
    return Rank(RankTier(badge // 10), RankDivision(badge % 10))


@dataclass(frozen=True)
class HeroCohort:
    minimum_badge: int
    maximum_badge: int
    rank_expansion: str
    expansion_history: tuple[dict[str, object], ...]

    @property
    def rank_range(self) -> RankRange:
        return RankRange(badge_rank(self.minimum_badge), badge_rank(self.maximum_badge))

    def as_dict(self) -> dict[str, object]:
        result = asdict(self)
        result["expansion_history"] = list(self.expansion_history)
        return result

    @classmethod
    def parse(cls, value: object) -> HeroCohort:
        row = object_dict(value)
        history = object_rows(row.get("expansion_history")) if row else None
        if row is None or not history:
            raise ArtifactError(
                "Hero has no rank expansion history; run refresh-evidence"
            )
        minimum = _required_int(row.get("minimum_badge"), "effective minimum badge")
        maximum = _required_int(row.get("maximum_badge"), "effective maximum badge")
        mode = str(row.get("rank_expansion"))
        start = _required_int(history[0].get("minimum_badge"), "starting minimum badge")
        try:
            cutoffs = ranked_cutoffs(start, maximum, mode)
        except ValueError as error:
            raise ArtifactError(f"Hero has an invalid rank range: {error}") from error
        if len(history) > len(cutoffs) or minimum != cutoffs[len(history) - 1]:
            raise ArtifactError("Hero has inconsistent effective ranks")
        for index, attempt in enumerate(history):
            _validate_attempt(
                attempt, cutoffs[index], maximum, last=index == len(history) - 1
            )
        return cls(minimum, maximum, mode, tuple(history))


def _validate_attempt(
    attempt: dict[str, object], minimum: int, maximum: int, *, last: bool
) -> None:
    if (
        attempt.get("minimum_badge") != minimum
        or attempt.get("maximum_badge") != maximum
    ):
        raise ArtifactError(
            "Hero rank expansion skipped a tier or changed the upper rank"
        )
    for key in (
        "discovery_rows",
        "selection_rows",
        "candidate_count",
        "discovery_owners",
        "selection_owners",
    ):
        _required_int(attempt.get(key), key)
    builds = _required_int(attempt.get("supported_builds"), "supported builds")
    if bool(builds) != last:
        raise ArtifactError("Hero expanded after support or has no supported build")
    if attempt.get("reason") != (
        "supported build available" if last else "no supported legal path"
    ):
        raise ArtifactError("Hero rank expansion has an invalid reason")

File: crates/deadlock-guides/src/hero_cohort.rs
use deadlock_data::{Error, Rank, RankRange, Result, array, object};
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::evidence_values::{integer, nonempty_text};

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum RankExpansion {
    Auto,
    Off,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct HeroCohortContent {
    pub minimum_badge: Rank,
    pub maximum_badge: Rank,
    pub rank_expansion: RankExpansion,
    pub expansion_history: Vec<Value>,
}

#[derive(Clone, Debug)]
pub struct HeroCohort(HeroCohortContent);

impl HeroCohort {
    /// # Errors
    /// Returns an error when the rank history has missing steps or continues after a supported build.
    pub fn from_document(value: &Value) -> Result<Self> {
        let content: HeroCohortContent = serde_json::from_value(value.clone())?;
        let history = array(&value["expansion_history"])?;
        let first = history.first().ok_or_else(|| {
            Error::new("Hero has no rank expansion history; run refresh-evidence")
        })?;
        let starting_badge = Rank::try_from(u16::try_from(integer(
            &first["minimum_badge"],
            "starting rank badge",
            1,
        )?)?)?;
        let cutoffs = calculate_rank_cutoffs(
            starting_badge,
            content.maximum_badge,
            content.rank_expansion,
        )?;
        if history.len() > cutoffs.len() || cutoffs[history.len() - 1] != content.minimum_badge {
            return Err(Error::new("Hero has inconsistent effective ranks"));
        }
        for (index, attempt) in history.iter().enumerate() {
            validate_attempt(
                attempt,
                cutoffs[index],
                content.maximum_badge,
                index + 1 == history.len(),
            )?;
        }
        Ok(Self(content))
    }

    #[must_use]
    pub const fn content(&self) -> &HeroCohortContent {
        &self.0
    }

    /// # Errors
    /// Returns an error when JSON serialization fails.
    pub fn to_document(&self) -> Result<Value> {
        Ok(serde_json::to_value(&self.0)?)
    }

    /// # Errors
    /// Returns an error when the rank range is invalid.
    pub fn rank_range(&self) -> Result<RankRange> {
        RankRange {
            minimum: self.0.minimum_badge,
            maximum: self.0.maximum_badge,
        }
        .validate()
    }
}

/// # Errors
/// Returns an error when the minimum rank exceeds the maximum rank.
pub fn calculate_rank_cutoffs(
    minimum: Rank,
    maximum: Rank,
    mode: RankExpansion,
) -> Result<Vec<Rank>> {
    RankRange { minimum, maximum }.validate()?;
    let mut cutoffs = vec![minimum];
    if mode == RankExpansion::Off || minimum.badge() == 11 {
        return Ok(cutoffs);
    }
    for tier in (1..minimum.tier()).rev() {
        cutoffs.push(Rank::try_from(tier * 10 + 1)?);
    }
    if cutoffs.len() == 1 {
        cutoffs.push(Rank::try_from(11)?);
    }
    Ok(cutoffs)
}

fn validate_attempt(attempt: &Value, minimum: Rank, maximum: Rank, last: bool) -> Result<()> {
    object(attempt)?;
    if integer(&attempt["minimum_badge"], "attempt minimum rank", 1)? != u64::from(minimum.badge())
        || integer(&attempt["maximum_badge"], "attempt maximum rank", 1)?
            != u64::from(maximum.badge())
    {
        return Err(Error::new(
            "Hero rank expansion skipped a tier or changed its maximum rank",
        ));
    }
    for key in [
        "discovery_rows",
        "selection_rows",
        "candidate_count",
        "discovery_owners",
        "selection_owners",
    ] {
        integer(&attempt[key], key, 0)?;
    }
    let builds = integer(&attempt["supported_builds"], "supported builds", 0)?;
    if (builds > 0) != last {
        return Err(Error::new(
            "Hero expanded after support or has no supported build",
        ));
    }
    let reason = if last {
        "supported build available"
    } else {
        "no supported legal path"
    };
    if nonempty_text(&attempt["reason"], "rank expansion reason")? != reason {
        return Err(Error::new("Hero rank expansion has an invalid reason"));
    }
    Ok(())
}
